modules/tinymce/src/core/main/ts/mode/Readonly.ts

```diff
@@ -156,14 +176,33 @@ const processReadonlyEvents = (editor: Editor, e: Event): void => {
 };
 
 const registerReadOnlySelectionBlockers = (editor: Editor): void => {
+  editor.on('beforeinput paste', (e) => {
+    if (isReadOnly(editor) && editor.mode.isSelectionEnabled()) {
+      e.preventDefault();
+    }
+  });
+
+  editor.on('BeforeSetContent', (e) => {
+    if (isReadOnly(editor) && editor.mode.isSelectionEnabled() && !e.initial) {
+      e.preventDefault();
+    }
+  });
+
+  const selectionModeCommands = Options.getSelectionModeAllowedCommands(editor);
+  editor.on('BeforeExecCommand', (e) => {
+    if (isReadOnly(editor) && editor.mode.isSelectionEnabled() && !Arr.contains(selectionModeCommands, e.command)) {
+      e.preventDefault();
+    }
+  });
+
   editor.on('ShowCaret', (e) => {
-    if (isReadOnly(editor)) {
+    if (isReadOnly(editor) && !editor.mode.isSelectionEnabled()) {
       e.preventDefault();
     }
   });
 
   editor.on('ObjectSelected', (e) => {
```

Comment:
For a fractional saving in bundle size, identical listeners can be registered with a single function. This was true before your change, but it's a nice improvement to include.
`editor.on('ShowCaret ObjectSelected', (e) => {`